### PR TITLE
Setup mypy

### PR DESCRIPTION
**Pull request by mattcurts**
config file is currently only set up to ignore missing imports as a means to work around coloredlogs error (will be fixed once merged) and aioredis error which will be addressed later once it is updated to <= 2.0.0.

### CODE REVIEW

File: mypy.ini [hunk 1]
```diff
@@ -0,0 +1,2 @@
+[mypy]
+ignore_missing_imports = True
```
Comment (igboyes) [inline mypy.ini:2]: Why was this necessary?

File: pyproject.toml [hunk 2]
```diff
@@ -24,8 +24,10 @@ psutil = "^5.8.0"
 pydantic = "^1.8.2"
 coloredlogs = "^15.0.1"
 motor = "^3.1.2"
-email-validator = "^1.2.1"
 enum-tools = "^0.9.0.post1"
+mypy = "^1.5.1"
```
Comment (igboyes) [inline pyproject.toml:28]: Install as dev dependency:
```
poetry add mypy --group dev
```

File: virtool_core/bio.py [hunk 3]
```diff
@@ -122,7 +122,7 @@ def read_fasta(path: Path) -> List[tuple]:
     return data
 
 
-async def read_fastq(f) -> Generator[tuple, None, list]:
+async def read_fastq(f) -> typing.AsyncGenerator[tuple, None]:
```
Comment (igboyes) [inline virtool_core/bio.py:125]:
```suggestion
async def read_fastq(f) -> AsyncGenerator[tuple, None]:
```

File: virtool_core/utils.py [hunk 4]
```diff
@@ -68,7 +83,9 @@ def compress_file_with_pigz(path: Path, target: Path, processes: int):
     :param target: path where the compressed file should be stored
     :param processes: number of processes allowable for pigz (-p argument)
     """
-    command = ["pigz", "-p", str(processes), "-k", "--stdout", path]
+    processes_str: str = str(processes)
+    abs_path: str = os.path.abspath(path)
+    command = ["pigz", "-p", processes_str, "-k", "--stdout", abs_path]
```
Comment (igboyes) [inline virtool_core/utils.py:88]: Use `pathlib` features.

https://stackoverflow.com/a/44569249
```suggestion
    command = ["pigz", "-p", processes_str, "-k", "--stdout", path.resolve()]
```

File: virtool_core/utils.py [hunk 5]
```diff
@@ -68,7 +83,9 @@ def compress_file_with_pigz(path: Path, target: Path, processes: int):
     :param target: path where the compressed file should be stored
     :param processes: number of processes allowable for pigz (-p argument)
     """
-    command = ["pigz", "-p", str(processes), "-k", "--stdout", path]
+    processes_str: str = str(processes)
```
Comment (igboyes) [inline virtool_core/utils.py:86]: 1. Don't use Hungarian notation.
2. Extra variable less clear than just putting type convertion in declaration of `command` list.
```
command = ["pigz", "-p", str(processes), ...]
```

File: virtool_core/utils.py [hunk 6]
```diff
@@ -111,7 +128,8 @@ def decompress_file_with_pigz(path: Path, target: Path, processes: int):
     :param target: path for the newly decompressed file to be stored
     :param processes: the number of allowable processes for pigz (-p argument)
     """
-    command = ["pigz", "-p", str(processes), "-d", "-k", "--stdout", path]
+    abs_path = os.path.abspath(path)
+    command = ["pigz", "-p", str(processes), "-d", "-k", "--stdout", abs_path]
```
Comment (igboyes) [inline virtool_core/utils.py:132]: Same as previous comment on `pathlib`. However, you did convert type inline here!

**Review by igboyes — CHANGES_REQUESTED** [review]
* For `typing` imports, use `from`. We should all know what these names are without the namespacing.
* Sorry, another deprectation is coming down the Python pipe. They are favouring typing like this `dict[str, int]` instead of `typing.Dict[str, int]`. May as well add this onto your PR.
* You have change to `|` over `Union` and `| None` instead of `Optional` in many places. Be consistent and make all of the changes. There are still lingering `Optional` and `Union` in some spots you touched.

**Comment by igboyes** [comment]
The Black formatter is also not running on this PR. The indentation breaks the style in some parts:
```
@asynccontextmanager
async def buffered_bulk_writer(
        collection, batch_size=100, session: Optional[AsyncIOMotorClientSession] = None
):
```

File: mypy.ini (hunk 1 above)
Comment (mattcurts) [reply mypy.ini:2]: This is currently to work around coloredlogs error (will be fixed once merged) and aioredis error which will be addressed later once it is updated to <= 2.0.0.

**Review by igboyes — APPROVED** [review]

**Pull request merged**